src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -2701,6 +2711,155 @@ void Lowering::TryLowerCselToCinc(GenTreeOp* select, GenTree* cond)
         }
     }
 }
+
+//----------------------------------------------------------------------------------------------
+// TryLowerAddSubToCombinedMulOp: Attempt to convert ADD and SUB nodes to a combined multiply
+// and add/sub operation, extending. Conversion can only happen if the operands to the
+// operation meet the following criteria:
+// - One op is a MUL_LONG containing two integer operands, and the other is a long.
+//
+// Arguments:
+//     op - The ADD or SUB node to attempt an optimisation on.
+//
+// Returns:
+//     A pointer to the next node to evaluate. On no operation, returns nullptr.
+//
+GenTree* Lowering::TryLowerAddSubToMulLongOp(GenTreeOp* op)
+{
+    assert(op->OperIs(GT_ADD, GT_SUB));
+
+    if (!comp->opts.OptimizationEnabled())
+        return nullptr;
+
+    if (op->isContained())
+        return nullptr;
+
+    if (!varTypeIsIntegral(op))
+        return nullptr;
+
+    if (op->gtFlags & GTF_SET_FLAGS)
+        return nullptr;
+
+    if (op->gtOverflow())
+        return nullptr;
+
+    GenTree* op1 = op->gtGetOp1();
+    GenTree* op2 = op->gtGetOp2();
+
+    // Select which operation is the MUL_LONG and which is the add value.
+    GenTreeOp* mul;
+    GenTree*   addVal;
+    if (op1->OperIs(GT_MUL_LONG))
+    {
+        // For subtractions, the multiply must be second, as [u/s]msubl performs:
+        // addValue - (mulValue1 * mulValue2)
+        if (op->OperIs(GT_SUB))
+        {
+            return nullptr;
+        }
+
+        mul    = op1->AsOp();
+        addVal = op2;
+    }
+    else if (op2->OperIs(GT_MUL_LONG))
+    {
+        mul    = op2->AsOp();
+        addVal = op1;
+    }
+    else
+    {
+        // Exit if neither operation are GT_MUL_LONG.
+        return nullptr;
+    }
+
+    // Additional value must be of long size.
+    // The compiler will check the relevant signage for us, so we don't need to worry about that.
+    if (!addVal->TypeIs(TYP_LONG))
+        return nullptr;
+
+    // Mul values must both be integers.
+    if (!genActualTypeIsInt(mul->gtOp1) || !genActualTypeIsInt(mul->gtOp2))
+        return nullptr;
+
+    // Create the new node and replace the original.
+    {
+        NamedIntrinsic intrinsicId =
+            op->OperIs(GT_ADD) ? NI_ArmBase_Arm64_MultiplyLongAdd : NI_ArmBase_Arm64_MultiplyLongSub;
+        GenTreeHWIntrinsic* outOp = comp->gtNewScalarHWIntrinsicNode(mul->IsUnsigned() ? TYP_ULONG : TYP_LONG,
```

Comment:
> mul->IsUnsigned() ? TYP_ULONG : TYP_LONG

It is illegal to type nodes as `TYP_ULONG`.

Reply:
> add an assert to make it self-explanatory?

Definitely - there is a long-term desire to implement an "IR verifier" component (we already have a lot of the pieces, but they're a bit ad-hoc) that would check for things like that. It just hasn't been done yet.

Reply:
Thank you, wasn't aware of this. I've fixed it as `TYP_LONG` and used the intrinsic's base type to determine unsigned/signed.